### PR TITLE
rtdl: implement DT_PREINIT_ARRAY

### PR DESCRIPTION
**Pull request by 64**
Follows on from #630.

### CODE REVIEW

**Review by Geertiebear — APPROVED** [review]
LGTM.

File: options/rtdl/generic/linker.cpp [hunk 1]
```diff
@@ -1293,9 +1310,19 @@ void Loader::_buildTlsMaps() {
 	}
 }
 
-void Loader::initObjects() {
+void Loader::initObjects(SharedObject *mainExecutable) {
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:1313]: Passing the main executable to this function is a bit ad-hoc. Can't we just store it in the `Loader` class instead?

File: options/rtdl/generic/linker.cpp [hunk 2]
```diff
@@ -1295,6 +1312,16 @@ void Loader::_buildTlsMaps() {
...
 
+	if (_mainExecutable && _mainExecutable->preInitArray) {
+		if (verbose)
+			mlibc::infoLogger() << "rtdl: Running DT_PREINIT_ARRAY functions" << frg::endlog;
+
+		__ensure(_mainExecutable->isMainObject);
+		__ensure((_mainExecutable->preInitArraySize % sizeof(InitFuncPtr)) == 0);
+		for(size_t i = 0; i < _mainExecutable->preInitArraySize / sizeof(InitFuncPtr); i++)
+			_mainExecutable->preInitArray[i]();
+	}
```
[the remark this answers is not in the document]
Comment (64) [reply options/rtdl/generic/linker.cpp:1323]: That shouldn't be necessary. But maybe we should ensure `!mainExecutable->isObjectInitialized` before running the pre-initializers; that'd have a similar effect.

**Review by Dennisbonke — APPROVED** [review]
LGTM

File: options/rtdl/generic/linker.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Geertiebear) [reply options/rtdl/generic/linker.cpp:1323]: Can we add that? Then this PR is good to go I think.

**Review by Geertiebear — APPROVED** [review]

**Pull request merged**